File: TheSky/JavaScript/LZ/scenes/MainScene/Adventure/Boss/bossChallengeLayer.js
var bossChallengeLayer = cc.Layer.extend({
	ctor:function(){
		this._super();
		cc.BuilderReader.registerController("BossChallengeOwner", {});
		this.node = cc.BuilderReader.load(ccbi_res.BossChallengeView_ccbi, this);
		if (this.node != null) {
			this.addChild(this.node);
		}
		this.autoTime = 2;
		this.logs = [];
		this.progress;
		this.refresh();
		common.swallowLayer(this, true);
	},
	refresh:function(){
		var bossData = bossdata.boss;
		var bossDic = BossModule.getBossAttrConfig(bossData.key);
		var boss = HeroModule.getHeroBust1ById(bossDic.npc[1].ID);
		if (boss) {
			this.bossBust.visible = true;
			this.bossBust.setTexture(boss);
			olAni.addPartical({plist : "images/eff_bossBust.plist", node : this.bossBust,
				pos : cc.p(this.bossBust.getContentSize().width / 2, this.bossBust.getContentSize().height / 2),
				 z : -1, tag : 100, scaleX : 1 / retina, scaleY: 1 / retina});
		}
		if (this.progress) {
			this.progress.removeFromParent(true);
			this.progress = null;
		}
		this.progress = new cc.ProgressTimer(new cc.Sprite("images/boss_hp.png"));
		this.progress.type = cc.ProgressTimer.TYPE_BAR;
		this.progress.midPoint = cc.p(0, 0);
		this.progress.barChangeRate = cc.p(1, 0);
		this.hpLayer.addChild(this.progress, 0, 101);
		this.progress.setPosition(cc.p(this.progress.getContentSize().width / 2, this.progress.getContentSize().height / 2));
		this.progress.percentage = Math.min(bossData.hpNow / bossData.hp * 100, 100)
		this.updateTimer();
	},
	closeItemClick:function(){
		this.removeFromParent(true);
	},
	updateTimer:function(){
		var playerData = bossdata.boss.playerData;
		var resurrectCD = 0;
		if (playerData && playerData[PlayerModule.getPlayerId()]) {
			resurrectCD = Math.max(0, playerData[PlayerModule.getPlayerId()].lastTime + playerData[PlayerModule.getPlayerId()].cdTime - Global.serverTime);
		}
		if (getUDBool(UDefKey.Setting_BossAuto) == "bossAuto" && resurrectCD != 0) {
			this.resurrentLabel.visible = true;
			this.resurrentLabel.setString(common.LocalizedString("boss_resurrect",DateUtil.second2hms(resurrectCD)));
		} else {
			this.resurrentLabel.visible = false;
		}
		var endCD = Math.max(0, bossdata.boss.endTime - Global.serverTime);
		this.endLabel.setString(common.LocalizedString("boss_end", DateUtil.second2hms(endCD)));
		if (endCD == 0) {
			postNotifcation(NOTI.BOSS_CLOSE_CHALLENGE);
			this.removeFromParent();
		}
		if (getUDBool(UDefKey.Setting_BossAuto) == "bossAuto") {
			this.challengeItem.enabled = false;
			this.resurrectItem.setEnabled(false);
			this.buffItem.setEnabled(false);
			this.challengeIcon.setSpriteFrame(cc.spriteFrameCache.getSpriteFrame("tiaozhanzhong_text.png"));
		} else {
			this.challengeItem.setEnabled(true);
			this.challengeIcon.setSpriteFrame(cc.spriteFrameCache.getSpriteFrame("boss_tiaozhan_text.png"));
			if (resurrectCD > 0) {
				this.resurrectItem.setEnabled(false);
				this.buffItem.setEnabled(true);
			} else {
				this.resurrectItem.setEnabled(true);
				this.buffItem.setEnabled(false);
			}
		}
		this.readLog();
	},
	removeNode:function(node){
		node.removeFromParent(true);
	},
	readLog:function(){
		var log = BossModule.popLog();
		if (log) {
			this.logs.push(log);
			var f = cc.spriteFrameCache.getSpriteFrame(HeroModule.getHeroHeadById(log.heroId));
			if (f) {
				var sp = new cc.Sprite(f)//cc.Sprite.createWithSpriteFrameName(f);
				this.infoBg.addChild(sp, 2);
				
				var x = this.infoBg.getContentSize().width / 2 * (1 + (Math.random() - 0.5) * 2);
				var y = this.infoBg.getContentSize().height / 2 * (1 + (Math.random() - 0.5) * 2);
				sp.setPosition(cc.p(x, y));
				var move = cc.MoveTo.create(0.2, cc.p(this.bossBust.x, this.bossBust.y + this.bossBust.getContentSize().height / 5));
				var fadeIn = cc.FadeTo.create(0.2, 255);
				var spawn = cc.Spawn.create(move, fadeIn);
				var array = [];
				array.push(spawn);
				array.push(cc.FadeTo.create(0.2, 0));
				array.push(cc.DelayTime.create(0.5));
				array.push(cc.CallFunc.create(this.removeNode(sp)));
				sp.runAction(cc.Sequence.create(array));
			}
			if (this.bossBust) {
				var oriPosX = this.bossBust.x;
				var oriPosY = this.bossBust.y;
				var array = [];
				array.push(cc.DelayTime.create(0.2));
				array.push(cc.MoveBy.create(0.05, cc.p(-12, 13)));
				array.push(cc.MoveBy.create(0.05, cc.p(14, -15)));
				array.push(cc.MoveBy.create(0.05, cc.p(13, -13)));
				array.push(cc.MoveBy.create(0.05, cc.p(-15, 0)));
				array.push(cc.MoveBy.create(0.05, cc.p(0, 15)));
				array.push(cc.MoveBy.create(0.05, cc.p(-12, -12)));
				array.push(cc.MoveBy.create(0.05, cc.p(13, -13)));
				array.push(cc.MoveBy.create(0.05, cc.p(-11, 13)));
				array.push(cc.MoveBy.create(0.05, cc.p(0, 11)));
				array.push(cc.MoveTo.create(0.05, cc.p(oriPosX, oriPosY)));
				this.bossBust.runAction(cc.Sequence.create(array));
			}
			if (log.damage && log.damage > 0) {
				var font = cc.LabelTTF.create("-"+log.damage,"images/fight_Red.fnt");
				font.setPosition(cc.p(this.bossBust.x, this.bossBust.y));
				this.infoBg.addChild(font, 2);
				font.setScale(2.5);
				var array = [];
				array.push(cc.DelayTime.create(0.2));
				array.push(cc.FadeTo.create(0.05, 255));
				array.push(cc.DelayTime.create(0.5));
				var move = cc.MoveBy.create(0.1, cc.p(0, 100));
				var fadeOut = cc.FadeTo.create(0.1, 0);
				var spawn = cc.Spawn.create(move, fadeOut);
				array.push(spawn);
				array.push(cc.DelayTime.create(0.5));
				array.push(cc.CallFunc.create(this.removeNode(font)));
				font.runAction(cc.Sequence.create(array));
			}
			if (this.logs.length > 4) {
				this.logs.pop();
			}
			for (var i = 0; i < this.logs.length; i++) {
				var logLabel = this["log" + (i+1)];
				logLabel.setString(common.LocalizedString("boss_log",this.logs[i].name,this.logs[i].damage));
				logLabel.visible = true;
			}
		}
	},
	autoFight:function(){
		if (getUDBool(UDefKey.Setting_BossAuto) == "bossAuto") {
			this.autoTime -= 1;
			if (this.autoTime == 0) {
				if (PlayerModule.getGold() < BossModule.getBossChallengeConfig("3").YXFZbuff) {
					common.showTipText(common.LocalizedString("ERR_1101"));
					this.autoTime = 2;
					setUDBool(UDefKey.Setting_BossAuto, "");
					return;
				}
				//TODO this.fight();
				BossModule.doOnBossBattle(3, this.challengeCallback.bind(this));
			}
		}
	},
	fight:function(){
		//TODO 加载BattleField数据
	},
	/**
	 * 立即复活
	 */
	resurrectItemClick:function(){
		BossModule.doOnBossBattle(2, this.challengeCallback.bind(this))
	},
	challengeCallback:function(){
		//this.logs = [{"damage":27161,"name":"201511191114","heroId":"hero_000402"}];
		//TODO 播放动画
	},
	/**
	 * 挑战
	 */
	challengeItemClick:function(){
		var resurrectCD = 0;
		var playerData = bossdata.boss.playerData;
		if (playerData && playerData[PlayerModule.getPlayerId()]) {
			resurrectCD = Math.max(0, playerData[PlayerModule.getPlayerId()].lastTime + playerData[PlayerModule.getPlayerId()].cdTime - Global.serverTime);
		}
		if (resurrectCD > 0) {
			common.showTipText(common.LocalizedString("boss_resurrect_tips"));
			return;
		}
		//TODO this.fight();
		BossModule.doOnBossBattle(1, this.challengeCallback.bind(this));
		this.refresh();
	},
	
	/**
	 * 浴血奋战
	 */
	buffItemClick:function(){
		olAni.addPartical({plist : "images/eff_bornOfFire.plist", node : this.buffItem,
			pos : cc.p(this.buffItem.getContentSize().width / 2, this.buffItem.getContentSize().height / 2),
			duration : 2, z : 100, tag : 100, scaleX : 1 / retina, scaleY: 1 / retina});
		var buff = BossModule.getBossChallengeConfig("3").YXFZbuff;
		//TODO this.fight(buff)
		BossModule.doOnBossBattle(3, this.challengeCallback.bind(this));
	},
	/**
	 * 自动挑战
	 */
	autoItemClick:function(){
		var text;
		if (getUDBool(UDefKey.Setting_BossAuto) == "bossAuto") {
			text = common.LocalizedString("boss_auto_tips_close");
		} else {
			text = common.LocalizedString("boss_auto_tips")
		}
		var cb = new ConfirmBox({info:text, confirm:function(){
			this.cardConfirmAction();
		}.bind(this)});
		cc.director.getRunningScene().addChild(cb);
	},
	cardConfirmAction:function(){
		if (getUDBool(UDefKey.Setting_BossAuto) == "bossAuto") {
			this.autoTime = 2;
			setUDBool(UDefKey.Setting_BossAuto, "");
			postNotifcation(NOTI.BOSS_CLOSE_CHALLENGE);
		} else {
			if (PlayerModule.getGold() >= BossModule.getBossChallengeConfig("3").YXFZcost) {
				setUDBool(UDefKey.Setting_BossAuto, "bossAuto");
				postNotifcation(NOTI.BOSS_CLOSE_CHALLENGE);
			} else {
				common.showTipText(common.LocalizedString("ERR_1101"));
				//TODO前往商店
			}
		}
	},
	onEnter:function(){
		this._super();
		var cache = cc.spriteFrameCache;
		cache.addSpriteFrames(common.formatLResPath("olRes/hero_head.plist"));
		addObserver(NOTI.BOSS_RESURRECT, "updateTimer", this);
		addObserver(NOTI.BOSS_AUTO_FIGHT, "autoFight", this);
	},
	onExit:function(){
		this._super();
		removeObserver(NOTI.BOSS_RESURRECT, "updateTimer", this);
		removeObserver(NOTI.BOSS_AUTO_FIGHT, "autoFight", this);
	},
});

File: TheSky/lua/op/AdventureView/bossChallengeLayer.lua
local _layer
local _priority = -134
local _autoTimer = 2
local _logs = {}
local _bNetwork
local progress

-- 名字不要重复
BossChallengeOwner = BossChallengeOwner or {}
ccb["BossChallengeOwner"] = BossChallengeOwner


local function closeItemClick()
    _layer:removeFromParentAndCleanup(true)
end
BossChallengeOwner["closeItemClick"] = closeItemClick


local function setMenuPriority()
    local menu = tolua.cast(BossChallengeOwner["menu"], "CCMenu")
    menu:setHandlerPriority(_priority - 1)
end

local function getLogCallback(url, rtnData)
    bossdata:insertLog(rtnData.info.logs)
    _bNetwork = false
end

local function getLog()
    if _bNetwork then
        return
    end
    _bNetwork = true
    if not bossdata.logTime or bossdata.logTime == 0 then
        bossdata.logTime = userdata.loginTime - 10
    end
    doActionNoLoadingFun("GET_BOSS_LOG", {bossdata.logTime}, getLogCallback)
    bossdata.logTime = userdata.loginTime
end

local function readLog()
    local log = bossdata:popLog()
    if log then
        table.insert(_logs, log)
        -- 飞蛾扑火
        local function removeNode(node)
            node:removeFromParentAndCleanup(true)
        end
        local bossBust = tolua.cast(BossChallengeOwner["bossBust"], "CCSprite")
        local infoBg = tolua.cast(BossChallengeOwner["infoBg"], "CCSprite")
        local f = CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(herodata:getHeroHeadByHeroId(log.heroId))
        if f then
            local sp = CCSprite:createWithSpriteFrame(f)
            infoBg:addChild(sp, 2)
            local x = infoBg:getContentSize().width / 2 * (1 + (RandomManager.fakeRandomRange(0, 1) - 0.5) * 2)
            local y = infoBg:getContentSize().height / 2 * (1 + (RandomManager.fakeRandomRange(0, 1) - 0.5) * 2)
            sp:setPosition(ccp(x, y))
            sp:setOpacity(0)
            local move = CCMoveTo:create(0.2, ccp(bossBust:getPositionX(), bossBust:getPositionY() + bossBust:getContentSize().height / 5))
            local fadeIn = CCFadeTo:create(0.2, 255)
            local spawn = CCSpawn:createWithTwoActions(move, fadeIn)
            local array = CCArray:create()
            array:addObject(spawn)
            array:addObject(CCFadeTo:create(0.2, 0))
            array:addObject(CCDelayTime:create(0.5))
            array:addObject(CCCallFuncN:create(removeNode))
            sp:runAction(CCSequence:create(array))
        end
        -- BOSS 虎躯一震
        if bossBust then
            local oriPosX, oriPosY = bossBust:getPosition()
            local array = CCArray:create()
            array:addObject(CCDelayTime:create(0.2))
            array:addObject(CCMoveBy:create(0.05, ccp(-12, 13)))
            array:addObject(CCMoveBy:create(0.05, ccp(14, -15)))
            array:addObject(CCMoveBy:create(0.05, ccp(13, -13)))
            array:addObject(CCMoveBy:create(0.05, ccp(-15, 0)))
            array:addObject(CCMoveBy:create(0.05, ccp(0, 15)))
            array:addObject(CCMoveBy:create(0.05, ccp(-12, -12)))
            array:addObject(CCMoveBy:create(0.05, ccp(13, -13)))
            array:addObject(CCMoveBy:create(0.05, ccp(-11, 13)))
            array:addObject(CCMoveBy:create(0.05, ccp(0, 11)))
            array:addObject(CCMoveTo:create(0.05, ccp(oriPosX, oriPosY)))
            bossBust:runAction(CCSequence:create(array))
        end
        -- BOSS 菊花一紧
        if log.damage and log.damage > 0 then
            local font = CCLabelBMFont:create("-"..log.damage, "images/fight_Red.fnt")
            font:setPosition(ccp(bossBust:getPositionX(), bossBust:getPositionY() + bossBust:getContentSize().height / 2))
            infoBg:addChild(font, 2)
            font:setOpacity(0)
            font:setScale(2.5)
            local array = CCArray:create()
            array:addObject(CCDelayTime:create(0.2))
            array:addObject(CCFadeTo:create(0.05, 255))
            array:addObject(CCDelayTime:create(0.5))
            local move = CCMoveBy:create(0.1, ccp(0, 100))
            local fadeOut = CCFadeTo:create(0.1, 0)
            local spawn = CCSpawn:createWithTwoActions(move, fadeOut)
            array:addObject(spawn)
            array:addObject(CCDelayTime:create(0.5))
            array:addObject(CCCallFuncN:create(removeNode))
            font:runAction(CCSequence:create(array))
        end
    end
    if #_logs > 4 then
        table.remove(_logs, 1)
    end
    for i,v in ipairs(_logs) do
        local logLabel = tolua.cast(BossChallengeOwner["log"..i], "CCLabelTTF")
        logLabel:setString(HLNSLocalizedString("boss.log", v.name, v.damage))
        logLabel:setVisible(true)
    end
end


local function _updateTimer()
    local resurrentLabel = tolua.cast(BossChallengeOwner["resurrentLabel"], "CCLabelTTF")
    local endLabel = tolua.cast(BossChallengeOwner["endLabel"], "CCLabelTTF")
    local playerData = bossdata.thisBoss.playerData
    local resurrectCD = 0
    if playerData and playerData[tostring(userdata.userId)] then
        resurrectCD = math.max(0, playerData[tostring(userdata.userId)].lastTime + playerData[tostring(userdata.userId)].cdTime - userdata.loginTime)
    end
    if not getUDBool(UDefKey.Setting_BossAuto, false) and resurrectCD ~= 0 then
        resurrentLabel:setVisible(true)
        resurrentLabel:setString(HLNSLocalizedString("boss.resurrect", DateUtil:second2hms(resurrectCD)))
    else
        resurrentLabel:setVisible(false)
    end
    local endCD = math.max(0, bossdata.thisBoss.endTime - userdata.loginTime)
    endLabel:setString(HLNSLocalizedString("boss.end", DateUtil:second2hms(endCD)))

    if endCD == 0 then
        getBossLayer():getBossInfo()
        _layer:removeFromParentAndCleanup(true)
        return
    end

    local challengeItem = tolua.cast(BossChallengeOwner["challengeItem"], "CCMenuItemImage")
    local resurrectItem = tolua.cast(BossChallengeOwner["resurrectItem"], "CCMenuItemImage")
    local buffItem = tolua.cast(BossChallengeOwner["buffItem"], "CCMenuItemImage")
    local autoItem = tolua.cast(BossChallengeOwner["autoItem"], "CCMenuItemImage")
    local challengeIcon = tolua.cast(BossChallengeOwner["challengeIcon"], "CCSprite")
    if getUDBool(UDefKey.Setting_BossAuto, false) then
        challengeItem:setEnabled(false)
        resurrectItem:setEnabled(false)
        buffItem:setEnabled(false)
        -- autoItem:setEnabled(false)
        challengeIcon:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("tiaozhanzhong_text.png"))
    else
        buffItem:setEnabled(true)
        challengeIcon:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("boss_tiaozhan_text.png"))
        if resurrectCD > 0 then
            challengeItem:setEnabled(false)
            resurrectItem:setEnabled(true)
        else
            challengeItem:setEnabled(true)
            resurrectItem:setEnabled(false)
        end
    end

    readLog()

end

local function _refresh()
    local bossDic = ConfigureStorage.bossAttr[bossdata.thisBoss.key]
    local bossBust = tolua.cast(BossChallengeOwner["bossBust"], "CCSprite")
    local texture = CCTextureCache:sharedTextureCache():addImage(herodata:getHeroBust1ByHeroId(bossDic.npc[1].ID))
    if texture then
        bossBust:setTexture(texture)
        -- renzhan
        HLAddParticleScale( "images/eff_bossBust.plist", BossChallengeOwner["bossBust"], ccp(BossChallengeOwner["bossBust"]:getContentSize().width / 2 , BossChallengeOwner["bossBust"]:getContentSize().height / 2), 5, -1, 100, 1/retina, 1/retina )
        -- 
    end

    -- local gold1 = tolua.cast(BossChallengeOwner["gold1"], "CCLabelTTF")
    -- gold1:setString(ConfigureStorage.bossPayAndBuff["2"].gold)
    -- local gold2 = tolua.cast(BossChallengeOwner["gold2"], "CCLabelTTF")
    -- gold2:setString(ConfigureStorage.bossPayAndBuff["3"].gold)


    local progressBg = tolua.cast(BossChallengeOwner["hpLayer"], "CCLayer")
    if progress then
        progress:removeFromParentAndCleanup(true)
        progress = nil
    end
    progress = CCProgressTimer:create(CCSprite:create("images/boss_hp.png"))
    progress:setType(kCCProgressTimerTypeBar)
    progress:setMidpoint(CCPointMake(0, 0))
    progress:setBarChangeRate(CCPointMake(1, 0))
    progress:setPosition(progress:getContentSize().width / 2, progress:getContentSize().height / 2)
    progressBg:addChild(progress,0, 101)
    progress:setPercentage(bossdata.thisBoss.hpNow / bossdata.thisBoss.hp * 100)
    _updateTimer()
end

local function _fight(buff)
    RandomManager.cursor = RandomManager.randomRange(0, 999)
    BattleField:bossFight(bossdata.thisBoss.key, bossdata.thisBoss.level, bossdata.thisBoss.hp, bossdata.thisBoss.hpNow, buff)
    BattleField.leftName = userdata.name
    BattleField.rightName = ConfigureStorage.bossAttr[bossdata.thisBoss.key].npc[1].name
end

local function _showFightResult()
    if getBossResultLayer() then
        getBossResultLayer():refresh()
    else
        getMainLayer():getParent():addChild(createBossResultLayer(-136), 100)
    end
end

local function challengeCallback(url, rtnData)
    runtimeCache.responseData = rtnData.info
    bossdata.thisBoss = rtnData.info.boss
    local logs = {
                    ["0"] = {
                                ["name"] = userdata.name, 
                                ["damage"] = BattleField.damage.left, 
                                ["heroId"] = herodata.heroes[herodata.form["0"]].heroId, 
                                ["time"] = userdata.loginTime
                            }
                }
    bossdata:insertLog(logs)
    CCDirector:sharedDirector():replaceScene(CCTransitionFade:create(0.6, FightingScene()))
end

local function errorCallback(url, rtnCode)
    if rtnCode == ErrorCodeTable.ERR_2901 or rtnCode == ErrorCodeTable.ERR_2902 then
        getBossLayer():getBossInfo()
        if getBossResultLayer() then
            getBossResultLayer():removeFromParentAndCleanup(true)
        end
        _layer:removeFromParentAndCleanup(true)
    end
end

local function challengeItemClick()
    Global:instance():TDGAonEventAndEventData("gorge3") 
    local resurrectCD = 0
    if playerData and playerData[tostring(userdata.userId)] then
        resurrectCD = math.max(0, playerData[tostring(userdata.userId)].lastTime + playerData[tostring(userdata.userId)].cdTime - userdata.loginTime)
    end
    if resurrectCD > 0 then
        ShowText(HLNSLocalizedString("boss.resurrect.tips"))
        return
    end
    _fight()
    doActionFun("BOSS_BATTLE", {BattleField.damage.left, 1}, challengeCallback, errorCallback) 
end
BossChallengeOwner["challengeItemClick"] = challengeItemClick

local function resurrectItemClick()
    Global:instance():TDGAonEventAndEventData("gorge1")
    _fight()
    doActionFun("BOSS_BATTLE", {BattleField.damage.left, 2}, challengeCallback, errorCallback)
end
BossChallengeOwner["resurrectItemClick"] = resurrectItemClick

local function buffItemClick()
    Global:instance():TDGAonEventAndEventData("gorge2")
    -- renzhan
    HLAddParticleScale( "images/eff_bornOfFire.plist", BossChallengeOwner["buffItem"], ccp(BossChallengeOwner["buffItem"]:getContentSize().width / 2 , BossChallengeOwner["buffItem"]:getContentSize().height / 2), 5, 100, 100, 1/retina, 1/retina )
    -- 
    _fight(ConfigureStorage.bossPayAndBuff["3"].buff)
    doActionFun("BOSS_BATTLE", {BattleField.damage.left, 3}, challengeCallback, errorCallback)
end
BossChallengeOwner["buffItemClick"] = buffItemClick

local function autoFightCallback(url, rtnData)
    runtimeCache.responseData = rtnData.info
    bossdata.thisBoss = rtnData.info.boss
    _autoTimer = 8
    progress:setPercentage(bossdata.thisBoss.hpNow / bossdata.thisBoss.hp * 100)
    local logs = {
                    ["0"] = {
                                ["name"] = userdata.name, 
                                ["damage"] = BattleField.damage.left, 
                                ["heroId"] = herodata.heroes[herodata.form["0"]].heroId, 
                                ["time"] = userdata.loginTime
                            }
                }
    bossdata:insertLog(logs)
    -- _showFightResult()
    local text = HLNSLocalizedString("boss.result.damage", BattleField.damage.left)
    local gain = rtnData.info.gain
    if gain["silver"] then
        text = string.format("%s\r\n%s", text, HLNSLocalizedString("boss.result.silver", gain["silver"]))
    end
    local itemId = "item_006"
    if gain["items"] and gain["items"][itemId] then
        local conf = wareHouseData:getItemResource(itemId)
        text = string.format("%s\r\n%s", text, HLNSLocalizedString("boss.result.train", gain["items"][itemId], conf.name))
    end
    if gain["encounter"] then
        print("--------- encounter -----------")
        text = string.format("%s\r\n%s", text, HLNSLocalizedString("boss.result.instruct"))
    end
    ShowText(text)
    -- renzhan
    HLAddParticleScale( "images/eff_bornOfFire.plist", BossChallengeOwner["buffItem"], ccp(BossChallengeOwner["buffItem"]:getContentSize().width / 2 , BossChallengeOwner["buffItem"]:getContentSize().height / 2), 5, 100, 100, 1/retina, 1/retina )
    -- 
end

local function _autoFight()
    if getUDBool(UDefKey.Setting_BossAuto, false) then
        _autoTimer = _autoTimer - 1
        if _autoTimer == 0 then
            if userdata.gold < ConfigureStorage.bossPayAndBuff["3"].gold then
                ShowText(HLNSLocalizedString("ERR_1101"))
                _autoTimer = 2
                setUDBool(UDefKey.Setting_BossAuto, false)
                return
            end
            _fight(ConfigureStorage.bossPayAndBuff["3"].buff)
            doActionFun("BOSS_BATTLE", {BattleField.damage.left, 3}, autoFightCallback, errorCallback)
        end
    end
end

local function cardConfirmAction()
    if getUDBool(UDefKey.Setting_BossAuto, false) then
        _autoTimer = 2
        setUDBool(UDefKey.Setting_BossAuto, not getUDBool(UDefKey.Setting_BossAuto, false))
        getBossLayer():refreshLayer()
    else
        if userdata.gold >= ConfigureStorage.bossPayAndBuff["3"].gold then
            setUDBool(UDefKey.Setting_BossAuto, not getUDBool(UDefKey.Setting_BossAuto, false))
            getBossLayer():refreshLayer()
        else
            ShowText(HLNSLocalizedString("ERR_1101"))
            getMainLayer():addChild(createShopRechargeLayer(-140), 100)
        end
    end
end

local function autoItemClick()
    Global:instance():TDGAonEventAndEventData("gorge4")
    local text
    if getUDBool(UDefKey.Setting_BossAuto, false) then
        text = HLNSLocalizedString("boss.auto.tips.close")
    else
        text = HLNSLocalizedString("boss.auto.tips")
    end
    getMainLayer():getParent():addChild(createSimpleConfirCardLayer(text), 100)
    SimpleConfirmCard.confirmMenuCallBackFun = cardConfirmAction
    SimpleConfirmCard.cancelMenuCallBackFun = nil
end
BossChallengeOwner["autoItemClick"] = autoItemClick

-- private方法如果没有上下调用关系可以写在createLayer外面
local function _init()
    -- get layer
    local  proxy = CCBProxy:create()
    local  node  = CCBReaderLoad("ccbResources/BossChallengeView.ccbi", proxy, true,"BossChallengeOwner")
    _layer = tolua.cast(node,"CCLayer")
    _refresh()
end

local function onTouchBegan(x, y)
    local touchLocation = _layer:convertToNodeSpace(ccp(x, y))
    local infoBg = tolua.cast(BossChallengeOwner["infoBg"], "CCSprite")
    local rect = infoBg:boundingBox()
    if not rect:containsPoint(touchLocation) then
        -- _layer:removeFromParentAndCleanup(true)
        return true
    end
    return true
end

local function onTouchEnded(x, y)

end

local function onTouch(eventType, x, y)
    if eventType == "began" then   
        return onTouchBegan(x, y)
    elseif eventType == "moved" then
    else
        return onTouchEnded(x, y)
    end
end

-- 该方法名字每个文件不要重复
function getBossChallengeLayer()
	return _layer
end

function createBossChallengeLayer(priority)
    _init()
    _priority = priority ~= nil and priority or -134

    function _layer:showFightResult()
        _showFightResult()
    end

    function _layer:close()
        closeItemClick()
    end

    local function _onEnter()
        print("onEnter")
        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/hero_head.plist")
        local seq = CCSequence:createWithTwoActions(CCDelayTime:create(0.2), CCCallFunc:create(setMenuPriority))
        _layer:runAction(seq)
        addObserver(NOTI_BOSS_RESURRECT, _updateTimer)
        addObserver(NOTI_BOSS_AUTO, _autoFight)
        addObserver(NOTI_BOSS_LOG, getLog)
        _autoTimer = 2
        _logs = {}
        _bNetwork = false
        getLog()
    end

    local function _onExit()
        print("onExit")
        _layer = nil
        _priority = -134
        _bNetwork = false
        progress = nil
        removeObserver(NOTI_BOSS_RESURRECT, _updateTimer)
        removeObserver(NOTI_BOSS_AUTO, _autoFight)
        removeObserver(NOTI_BOSS_LOG, getLog)
    end

    --onEnter onExit
    local function _layerEventHandler(eventType)
        if eventType == "enter" then
            if _onEnter then _onEnter() end
        elseif eventType == "exit" then
            if _onExit then _onExit() end
        end
    end


    _layer:registerScriptTouchHandler(onTouch ,false , _priority ,true )
    _layer:setTouchEnabled(true)
    _layer:registerScriptHandler(_layerEventHandler)

    return _layer
end